### ISSUE TITLE
Feature Request: StatefulHookWidget

### ISSUE THREAD

**Issue by esDotDev** [rrousselGit/flutter_hooks#126]
While the ability for hooks to reduce standard Flutter "controller" boilerplate are a massive win, and are desperately needed, **useState** does not offer any significant wins in terms of reducing boilerplate, in fact it tends to increase it substantially.

Would it be possible to create StatefulHookWidget, where we can still use setState, and easily share (non boilerplate) State, while using all the awesome useAnimation, useFuture, useStreamController, etc to strip away all that standard Flutter boilerplate?

It feels like the major benefits of the hook paradigm (DRY code, elimination of bugs), are being countered by the strict decision to not allow any state at all which forces all kinds of refactoring, for no apparent gain. The ability to directly share state between related widgets, and easily trigger rebuilds, allows the elimination of a lot of boilerplate and keeps things flexible during development when things are changing quickly.

Basically, the way I see it, the current Flutter implementation of Controllers/Tickers, is quite _poor_, and needs to be addressed. But their current version of state management, is fine. And actually much more flexible than this approach. 

It would be great to have access to the half of this we really need, without being saddled with the half we don't. 

Thoughts?

**Comment by rrousselGit** [comment]
>  But their current version of state management, is fine. And actually much more flexible than this approach.

Do you have some examples?
Without concrete example on why this would be beneficial, I will disagree with this statement

Hooks aren't just a simplification of "controller"s/"builder"s. They are also a paradigm shift to make the state:
- immutable
- computed (automatically invalidates any cache)
- declarative / top-down
- reusable

**Comment by esDotDev** [comment]
I get what _hooks_ are, but my argument is that I really don't see why we need entire paradigm in dart. It makes sense for Javasript, when they're building on a 20yr old language, and the entire concept of classes and functions is a mess. But we have a modern language, with modern features.

To clarify, I didn't argue it's _better_ or _worse_, but it's objectively more _flexible_ to use a state. It can be passed around, enabling a lot of use cases without boilerplate. Being able to call setState is also more flexible, as it does not impose the requirement that you have access to some ValueNotifier to trigger the rebuild, again just enabling a variety of use cases that become tougher without it. 

In terms of concrete example, it feels very obvious on it's face. Any time you have a large stateful widget, that you then want to break out in to sub widgets, you will get into parameter hell with this approach. Especially if you want to have a button handler defined back in the main class for better code organization. Rather than passing a  value to my view, and a handler function that calls set state, I now have to pass the entire value notifier, carry that along to the method handler, and then 

Consider the following example. Where exactly is the win, I don't see it:
http://screens.gskinner.com/shawn/Photoshop_2020-03-02_08-57-56.png

I think there's too much Javascript baggage being brought with this. If you think about it from a Flutter first POV, what we really _need_ in Flutter, is basically just self-contained mixins, mixins that can hold their own little state, and get setup and torn down automatically. Then we can have a stack of those in any widget, providing whatever re-usable functionality we need.

**Comment by rrousselGit** [comment]
Your problem is that you are trying to use _method_, but you shouldn't.


```dart
class Example with HookWidget {
  @override
  Widget build(BuildContext context) {
    final state = useState(0);
    
    void increment() {
      state.value++;
    }
    
    return RaisedButton(
      onPressed: increment,
      child: Text('+1'),
    );
  }
}
```

instead of:

```dart
class Example with HookWidget {
  @override
  Widget build(BuildContext context) {
    final state = useState(0);
    

    return RaisedButton(
      onPressed: () => increment(state),
      child: Text('+1'),
    );
  }

  void increment(ValueNotifier<int> state) {
    state.value++;
  }
}
```

**Comment by rrousselGit** [comment]
The problem is not so much the flexibility that it provides, because I can agree with the idea.
More choice is obviously more flexible than a single choice.

The issue is the maintenance cost.
Doing it is quite a lot of work and may prevent future optimizations

**Comment by esDotDev** [comment]
Ah, I see, declare the handlers within the scope of the function so they can just easily access state. That does help make less verbose. But how does that work if I need to have a dedicated view to build my very complex widget? I still need to pass in all of these functions since the view can not directly access them on the state, and we're back to needless boilerplate (albeit, less so).

Hopefully you can at least see my argument that, there would be some benefit in providing the 'micro states' power of hooks, with the well known and robust ability to have a central state object that can be used as normal.

**Comment by rrousselGit** [comment]
Well, the example you gave of passing the `State` to another widget is a bad idea.
It defeats the purpose of separating the state from the view.

You do not have to use one `useState` per property anyway. You can have classes:


```dart
class MyState with ChangeNotifier {
  String many;
  String properties;
}

T useMemoNotifier<T extends ChangeNotifier>(T Function() cb) {
  final res = useMemoized(cb);
  useEffect(() {
    return res.dispose;
  }, const []);

  useListenable(res);
  return res;
}

class Example extends HookWidget {
  @override
  Widget build(BuildContext context) {
    final state = useMemoNotifier(() => MyState());

    return MyView(state);
  }
}
```

**Comment by rrousselGit** [comment]
Personally, I use [freezed](https://github.com/rrousselGit/freezed)

So I would write:


```dart
@freezed
abstract class MyState with _$MyState {
  const factory MyState({@Default(0) int count, String many, String properties}) = _MyState;
}

class Example extends HookWidget {
  @override
  Widget build(BuildContext context) {
    final myState = useState(const MyState());

    void increment() {
      myState.value = myState.value.copyWith(count: myState.value + 1);
    }

    return MyView(state);
  }
}
```

**Comment by esDotDev** [comment]
It certainly does not defeat the purpose of separating the state from the view. The view contains all declarative code, and the state contains all state and business logic. This is excellent for scaling when your views and states can scale to several hundreds of lines of code. In this case, these 2 widgets are two sides of the same coin, I don't agree this is bad practice. 

The solution above seems like it would work well, but now all my simple data objects need to be change notifiers. And I mean, look at how much boilerplate is sprouting up, just to replace a simple .state and setState() calls. I think it's speaking for itself here... 

Freeze is very neat, but code generation is not a road we want to travel down in production, with multiple (novice) developers, and a tight timeline.

**Comment by esDotDev** [comment]
I would take a stab at implementing this myself, I just don't have enough knowledge about the flutter lifecycle. Seems like it can't be that hard to simply have a state, that can also call useFoo to get some microstate that has it's own setup/teardown methods. And then to me, that gets you 95% of what is truly helpful about Hooks.

**Comment by rrousselGit** [comment]
Closing in favor of #137

**Comment by esDotDev** [comment]
Nice!
